File: data_structures/linkedLists/lec2/dLL.cpp
#include <iostream>
#include <cmath>

using namespace std;

class Node
{
public:
    int data;
    Node *next;
    Node *prev;

public:
    Node(int data1)
    {
        data = data1;
        next = nullptr;
        prev = nullptr;
    }

    Node(int data1, Node *next1, Node *prev1)
    {
        data = data1;
        next = next1;
        prev = prev1;
    }
};

Node *deleteLastNode(Node *head)
{
    if (head == nullptr || head->next == nullptr)
    {
        return nullptr;
    }

    Node *tail = head;
    while (tail->next != nullptr)
    {
        tail = tail->next;
    }

    Node *newTail = tail->prev;
    newTail->next = nullptr;
    delete tail;

    return head;
}
// Class definition of a node ends here.
Node *insertAtHead(Node *head, int value)
{
    Node *newNode = new Node(value, head, nullptr);
    return newNode;
}

Node *insertAtTail(Node *head, int value)
{
    Node *newNode = new Node(value);
    if (head == nullptr)
    {
        return newNode;
    }

    Node *tail = head;
    while (tail->next != nullptr)
    {
        tail = tail->next;
    }

    tail->next = newNode;
    newNode->prev = tail;

    return head;
}

Node *convertArrToDLL(vector<int> arr1)
{
    Node *head = new Node(arr1[0]);

    Node *tail = head;

    for (int i = 1; i < arr1.size(); i++)
    {
        Node *newNode = new Node(arr1[i], nullptr, tail);
        tail->next = newNode;
        newNode->prev = tail;
        tail = tail->next;
    }

    return head;
}

Node *reverseDLL(Node *head)
{
    if (head == nullptr || head->next == nullptr)
    {
        return nullptr;
    }

    Node *prev = nullptr;
    Node *current = head;

    while (current != nullptr)
    {
        prev = current->back;
        current->back = current->next;
        current->next = prev;
        current = current->back;
    }

    return prev->back;
}

void printDll(Node *head)
{
    while (head != nullptr)
    {
        cout << head->data << " <-> ";
        head = head->next;
    }
}

int main()
{
    vector<int> arr = {12, 5, 8, 7, 4};
    Node *head = convertArrToDLL(arr);

    cout << "Doubly Linked List Initially: " << endl;
    printDll(head);

    cout << endl
         << "Doubly Linked List After Inserting at the tail with value 10: " << endl;
    // Insert a node with value 10 at the end
    head = insertAtTail(head, 10);
    head = insertAtHead(head, 10);
    head = deleteLastNode(head);
    printDll(head);

    return 0;
}

File: data_structures/ll_final/dll.py
class DNode: 
    def __init__(self, val, next=None, prev=None): 
        self.val = val 
        self.next = next
        self.prev = prev

def printDLL(head):
    while head is not None: 
        print(head.val, "<->", end=" ")
        head = head.next 

def insertAtEnd(head, k):
    newNode = DNode(k, None, None)

    if head is None: 
        return DNode(k, None, None) 
    
    temp = head
    while temp.next is not None: 
        temp = temp.next 

    temp.next = newNode
    newNode.prev = temp 
    
    return head 

def deleteNode(head): 
    # Deleting the tail of a linked list. 
    if head is None or head.next is None: 
        return None 
    
    temp = head 
    while temp.next is not None: 
        temp = temp.next 
    
    previousNode = temp.prev 
    previousNode.next = None 

    temp.prev = None
    del temp 

    return head 

def reverseDLL(head): 
    if head is None or head.next is None: 
        return head 
    
    prev = None 
    current = head 
    while current is not None: 
        prev = current.prev  
        current.prev = current.next 
        current.next = prev 
        current = current.prev 
    return prev.prev 

def convert_arr_to_dll(arr):
    # Create the head node with
    # the first element of the array
    head = DNode(arr[0])
    # Initialize 'prev' to the head node
    prev = head

    for i in range(1, len(arr)):
        # Create a new node with data from the
        # array and set its 'back' pointer
        # to the previous node
        temp = DNode(arr[i], None, prev)
        
        # Update the 'next' pointer of the
        # previous node to point to the new node
        prev.next = temp
        # Move 'prev' to the newly created 
        # node for the next iteration
        prev = temp

    # Return the head of the doubly linked list
    return head



# Example usage:
arr = [12, 5, 6, 8, 4]
# Convert the array to a
# doubly linked list
head = convert_arr_to_dll(arr)
# Print the doubly linked list
print('Doubly Linked List Initially:  ')
printDLL(head)

print('Doubly Linked List After Reversing :')

# Reverse the doubly linked list
head = reverseDLL(head)
# Print the reversed doubly linked list
printDLL(head)


    